README.rst
##############
[exif] Package
##############

.. image:: https://www.gitlab.com/TNThieding/exif/badges/master/pipeline.svg
    :target: https://gitlab.com/TNThieding/exif

.. image:: https://www.gitlab.com/tnthieding/exif/badges/master/coverage.svg
    :target: https://gitlab.com/TNThieding/exif

.. image:: https://readthedocs.org/projects/exif/badge/?version=latest
    :target: https://exif.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status

.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
    :target: https://github.com/psf/black

.. image:: http://www.mypy-lang.org/static/mypy_badge.svg
    :target: http://mypy-lang.org/
    :alt: Checked with mypy

Read and modify image EXIF metadata using Python without any third-party software
dependencies. For example, batch process image metadata using a Python script.

***********
Quick Start
***********

Open an image with EXIF metadata using the Python ``open()`` built-in function. Ensure the
binary mode flag is set. Pass this image file object into the ``exif.Image`` class::

    >>> from exif import Image
    >>> with open('grand_canyon.jpg', 'rb') as image_file:
    ...     my_image = Image(image_file)
    ...
    >>> my_image.has_exif
    True

List EXIF attributes using the ``list_all()`` method::

    >>> my_image.list_all()
    ['_exif_ifd_pointer', '_gps_ifd_pointer', 'aperture_value', 'brightness_value', 'color_space',
     'components_configuration', 'compression', 'datetime', 'datetime_digitized', 'datetime_original', 'exif_version',
     'exposure_bias_value', 'exposure_mode', 'exposure_program', 'exposure_time', 'f_number', 'flash',
     'flashpix_version', 'focal_length', 'focal_length_in_35mm_film', 'gps_altitude', 'gps_altitude_ref',
     'gps_datestamp', 'gps_dest_bearing', 'gps_dest_bearing_ref', 'gps_horizontal_positioning_error',
     'gps_img_direction', 'gps_img_direction_ref', 'gps_latitude', 'gps_latitude_ref', 'gps_longitude',
     'gps_longitude_ref', 'gps_speed', 'gps_speed_ref', 'gps_timestamp', 'jpeg_interchange_format',
     'jpeg_interchange_format_length', 'lens_make', 'lens_model', 'lens_specification', 'make', 'maker_note',
     'metering_mode', 'model', 'orientation', 'photographic_sensitivity', 'pixel_x_dimension', 'pixel_y_dimension',
     'resolution_unit', 'scene_capture_type', 'scene_type', 'sensing_method', 'shutter_speed_value', 'software',
     'subject_area', 'subsec_time_digitized', 'subsec_time_original', 'white_balance', 'x_resolution',
     'y_and_c_positioning', 'y_resolution']

Access EXIF metadata tags using Python attribute notation::

    >>> # Read tags with Python "get" notation.
    >>> my_image.gps_latitude
    (36.0, 3.0, 11.08)
    >>> my_image.gps_longitude
    (112.0, 5.0, 4.18)
    >>> my_image.model
    'iPhone 7'
    >>>
    >>> # Modify tags with Python "set" notation.
    >>> my_image.make = "Python"
    >>>
    >>> # Delete tags with Python "del" notation.
    >>> del my_image.gps_latitude
    >>> del my_image.gps_longitude
    >>>
    >>> # Add new tags with Python "set" notation.
    >>> from exif import LightSource
    >>> my_image.light_source = LightSource.DAYLIGHT

Write the image with modified EXIF metadata to an image file using ``open()`` in binary
write mode::

    >>> with open('modified_image.jpg', 'wb') as new_image_file:
    ...     new_image_file.write(my_image.get_file())
    ...

Refer to the `usage page <https://exif.readthedocs.io/en/latest/usage.html>`_ for information and examples of alternative ways to access EXIF tags (e.g.
with index/item syntax or with methods).


docs/release_notes.rst
#############
Release Notes
#############

*******************************************************
[1.6.0] Support ``rating`` and ``rating_percent`` tags.
*******************************************************

Now, users can access the ``rating`` and ``rating_percent`` EXIF tags.

This release contains changes submitted via GitLab merge request by the following user:

* Gerald Triendl (TriG81)


**************************************************************
[1.5.0] Support ``photographic_sensitivity`` tag. (2023-01-09)
**************************************************************

Now, users can read and write the ``photographic_sensitivity`` EXIF tag.

This release contains changes submitted via GitHub merge request by the following user:

* Luca Di Leo (HeDo88TH)


***********************************************************
[1.4.2] Improve data packing when adding tags. (2022-12-28)
***********************************************************

Previously, new tags added to an image were always placed at the end of the metadata section. Now, new tags are inserted
directly after the last tag value block and not just at the end of the metadata section. Therefore, the data is inserted
before the trailing unused bytes.

This patch contains changes submitted via GitLab merge request by the following user:

* Patrick (buergi)


**************************************************************************
[1.4.1] Fix logic for checking if a value fits in an IFD tag. (2022-12-23)
**************************************************************************

Previously, modifying EXIF attribute values unintentionally changed untouched attributes
(e.g., ``components_configuration`` and ``subject_area``) due to a bug within the helper
function ``value_fits_in_ifd_tag``.

This patch addresses the following GitLab user issue:

* Modifying EXIF changes untouched attributes. (https://gitlab.com/TNThieding/exif/-/issues/68)

This patch contains changes submitted via GitLab merge request by the following user:

* Patrick (buergi)


*************************************************************
[1.4.0] Late-November 2022 merge request rollup. (2022-11-25)
*************************************************************

This release includes community-contributed merge requests:

* Add ``focal_plane_x_resolution`` and ``focal_plane_y_resolution`` to the attribute mapping. (https://gitlab.com/TNThieding/exif/merge_requests/55)
* Fix incorrect determination of APP1 size. (https://gitlab.com/TNThieding/exif/-/merge_requests/56)
* Fix wrong offset of newly-created ITD. (https://gitlab.com/TNThieding/exif/-/merge_requests/59)
* Add support for multi-valued signed rationals. (https://gitlab.com/TNThieding/exif/-/merge_requests/60)

Thank you to the following users for their contributions:

* Patrick (buergi)
* simonschindler


*******************************************************************
[1.3.5] Support initial stable release of ``plum-py``. (2022-04-17)
*******************************************************************

Update package to support the initial stable release of its ``plum-py`` dependency.


************************************************************
[1.3.4] Decode Windows XP style tags as UTF-16. (2021-12-09)
************************************************************

Previously, the package decoded ASCII characters within Windows XP style tags. Now, the
package decodes Windows XP style tags as UTF-16.

This patch addresses the following GitLab user issue:

* ``xp_comment`` and other Windows XP tags don't handle Unicode strings correctly.
  (https://gitlab.com/TNThieding/exif/-/issues/53)


*******************************************************************
[1.3.3] Omit unknown values from ``get_all()`` method. (2021-11-07)
*******************************************************************

Previously, ``get_all()`` did not catch exceptions when reading each EXIF tag. If a single
attribute had an incorrectly-encoded value, the ``get_all()`` method raised an exception.
Now, the ``get_all()`` method catches exceptions due to unknown or unreadable values and
logs them as a warning.

This patch addresses the following GitLab user issue:

* ``ValueError: 0 is not a valid Orientation`` returned from the ``Image.get_all()``.
  method (https://gitlab.com/TNThieding/exif/-/issues/52)


**********************************************************
[1.3.2] Add support for writing various tags. (2021-09-04)
**********************************************************

Previously, attempting to add the following tags to an image raised an ``AttributeError``:

* Body Serial Number
* ISO Speed
* Lens Specification
* Lens Make
* Lens Model
* Lens Serial Number

This patch addresses the following GitLab user issue:

* Trouble setting tags. (https://gitlab.com/TNThieding/exif/-/issues/48)


*******************************************************************
[1.3.1] Fix value of ``SceneCaptureType.NIGHT_SCENE``. (2021-07-03)
*******************************************************************

Previously, ``SceneCaptureType.NIGHT_SCENE`` erroneously had a value of ``2``. Now, it has a value of ``3`` in
accordance with the EXIF specification.

This patch contains changes submitted via GitLab merge request by the following user:

* Alex Mykyta (amykyta3)


***********************************************************
[1.3.0] Consume latest version of ``plum-py``. (2021-06-13)
***********************************************************

Overhaul package internals to leverage ``plum-py`` version ``0.5.0`` and higher. Since the ``plum-py`` package only
supports Python 3.7 and higher, this release drops support for Python 3.6.


****************************************************
[1.2.2] Late-April 2021 bug fix rollup. (2021-04-23)
****************************************************

This patch addresses the following GitLab user issues:


exif/_constants.py
"""Package constants."""
# pylint: disable=too-few-public-methods

from enum import IntEnum


class ColorSpace(IntEnum):
    """Color space specifier."""

    SRGB = 1
    "sRBG"

    UNCALIBRATED = 0xFFFF
    "Uncalibrated or Other"


class ExifMarkers:
    """EXIF marker segments bytes."""

    SEG_PREFIX = b"\xff"
    """Generic Segment Prefix"""

    SOI = SEG_PREFIX + b"\xd8"
    """Start of Image"""

    APP1 = SEG_PREFIX + b"\xe1"
    """EXIF Attribute Information (Application Segment 1)"""

    APP2 = SEG_PREFIX + b"\xe2"
    """EXIF Extended Data (Application Segment 2)"""

    DQT = SEG_PREFIX + b"\xdb"
    """Quantization Table Definition"""

    DHT = SEG_PREFIX + b"\xc4"
    """Huffman Table Definition"""

    DRI = SEG_PREFIX + b"\xdd"
    """Restart Interoperability Definition"""

    SOF = SEG_PREFIX + b"\xc0"
    """Start of Frame"""

    SOS = SEG_PREFIX + b"\xda"
    """Start of Scan"""

    EOI = SEG_PREFIX + b"\xd9"
    """End of Image"""


class ExifTypes(IntEnum):
    """EXIF datatype indicator for IFD structure."""

    BYTE = 1
    """8-Bit Unsigned Integer"""

    ASCII = 2
    """Null-Terminated ASCII Codes"""

    SHORT = 3
    """16-Bit Unsigned Integer"""

    LONG = 4
    """32-Bit Unsigned Integer"""

    RATIONAL = 5
    """Two (Numerator and Denominator) LONGs"""

    SSHORT = 8
    """16-Bit Signed Integer"""

    SLONG = 9
    """32-Bit Signed Integer"""

    SRATIONAL = 10
    """Two (Numerator and Denominator) SLONGs"""


class ExposureMode(IntEnum):
    """Exposure mode set when the image was shot."""

    AUTO_EXPOSURE = 0
    """Auto Exposure"""

    MANUAL_EXPOSURE = 1
    """Manual Exposure"""

    AUTO_BRACKET = 2
    """Auto Bracket"""


class ExposureProgram(IntEnum):
    """Class of the program used by the camera to set exposure when the picture is taken."""

    NOT_DEFINED = 0
    """Not Defined"""

    MANUAL = 1
    """Manual"""

    NORMAL_PROGRAM = 2
    """Normal Program"""

    APERTURE_PRIORITY = 3
    """Aperture Priority"""

    SHUTTER_PRIORITY = 4
    """Shutter Priority"""

    CREATIVE_PROGRAM = 5
    """Creative Program (Biased Toward Depth of Field)"""

    ACTION_PROGRAM = 6
    """Action Program (Biased Toward Fast Shutter Speed)"""

    PORTRAIT_MODE = 7
    """Portrait Mode (For Closeup Photos with the Background out of Focus)"""

    LANDSCAPE_MODE = 8
    """Landscape Kode (For Landscape Photos with the Background in Focus)"""


class GpsAltitudeRef(IntEnum):
    """Altitude used as the reference altitude."""

    ABOVE_SEA_LEVEL = 0
    """Above Sea Level"""

    BELOW_SEA_LEVEL = 1
    """Below Sea Level"""


class LightSource(IntEnum):
    """Class of the program used by the camera to set exposure when the picture is taken."""

    UNKNOWN = 0
    """Unknown"""

    DAYLIGHT = 1
    """Daylight"""

    FLUORESCENT = 2
    """Fluorescent"""

    TUNGSTEN = 3
    """Tungsten (Incandescent Light)"""

    FLASH = 4
    """Flash"""

    FINE_WEATHER = 9
    """Fine Weather"""

    CLOUDY_WEATHER = 10
    """Cloudy Weather"""

    SHADE = 11
    """Shade"""

    DAYLIGHT_FLUORESCENT = 12
    """Daylight Fluorescent (D 5700 - 7100K)"""

    DAY_WHITE_FLUORESCENT = 13
    """Day White Fluorescent (N 4600 - 5400K)"""

    COOL_WHITE_FLUORESCENT = 14
    """Cool White Fluorescent (W 3900 - 4500K)"""

    WHITE_FLUORESCENT = 15
    """White Fluorescent (WW 3200 - 3700K)"""

    STANDARD_LIGHT_A = 17
    """Standard Light A"""

    STANDARD_LIGHT_B = 18
    """Standard Light B"""

    STANDARD_LIGHT_C = 19
    """Standard Light C"""

    D55 = 20
    """D55"""

    D65 = 21
    """D65"""

    D75 = 22
    """D75"""

    D50 = 23
    """D50"""

    ISO_STUDIO_TUNGSTEN = 24
    """ISO Studio Tungsten"""

    OTHER = 255
    """Other Light Source"""


class MeteringMode(IntEnum):
    """Metering mode."""

    UNKNOWN = 0
    """Unknown"""

    AVERAGE = 1
    """Average"""

    CENTER_WEIGHTED_AVERAGE = 2
    """Center Weighted Average"""

    SPOT = 3
    """Spot"""

    MULTI_SPOT = 4
    """Multi Spot"""

    PATTERN = 5
    """Pattern"""

    PARTIAL = 6
    """Partial"""

    OTHER = 255
    """Other"""


class Orientation(IntEnum):
    """Image orientation in terms of rows and columns."""

    TOP_LEFT = 1
    """The 0th row is at the visual top of the image and the 0th column is the visual
    left-hand side."""

    TOP_RIGHT = 2
    """The 0th row is at the visual top of the image and the 0th column is the visual
    right-hand side."""

    BOTTOM_RIGHT = 3
    """The 0th row is at the visual bottom of the image and the 0th column is the visual
    right-hand side."""

    BOTTOM_LEFT = 4
    """The 0th row is at the visual bottom of the image and the 0th column is the visual
    left-hand side."""

    LEFT_TOP = 5
    """The 0th row is the visual left-hand side of the image and the 0th column is the
    visual top."""

    RIGHT_TOP = 6
    """The 0th row is the visual right-hand side of the image and the 0th column is the
    visual bottom."""

    RIGHT_BOTTOM = 7
    """The 0th row is the visual right-hand side of the image and the 0th column is the
    visual bottom."""

    LEFT_BOTTOM = 8
    """The 0th row is the visual left-hand side of the image and the 0th column is the
    visual bottom."""


class ResolutionUnit(IntEnum):
    """Unit for measuring X resolution and Y resolution tags."""

    INCHES = 2
    """Inches or Unknown"""

    CENTIMETERS = 3
    """Centimeters"""


class Saturation(IntEnum):
    """Saturation processing applied by camera."""

    NORMAL = 0
    """Normal Saturation"""

    LOW = 1
    """Low Saturation"""

    HIGH = 2
    """High Saturation"""


class SceneCaptureType(IntEnum):
    """Type of scene that was shot or the mode in which the image was shot."""

    STANDARD = 0
    """Standard"""

    LANDSCAPE = 1
    """Landscape"""

    PORTRAIT = 2
    """Portrait"""

    NIGHT_SCENE = 2
    """Night Scene"""


class SensingMethod(IntEnum):
    """Image sensor type on the camera or input device."""

    NOT_DEFINED = 1
    """Not Defined"""

    ONE_CHIP_COLOR_AREA_SENSOR = 2
    """One-Chip Color Area Sensor"""

    TWO_CHIP_COLOR_AREA_SENSOR = 3
    """Two-Chip Color Area Sensor"""

    THREE_CHIP_COLOR_AREA_SENSOR = 4
    """Three-Chip Color Area Sensor"""

    COLOR_SEQUENTIAL_AREA_SENSOR = 5
    """Color Sequential Area Sensor"""

    TRILINEAR_SENSOR = 7
    """Trilinear Sensor"""

    COLOR_SEQUENTIAL_LINEAR_SENSOR = 8
    """Color Sequential Linear Sensor"""


class Sharpness(IntEnum):
    """Sharpness processing applied by camera."""

    NORMAL = 0
    """Normal"""

    SOFT = 1
    """Soft"""

    HARD = 2
    """Hard"""


class WhiteBalance(IntEnum):
    """White balance mode set when the image was shot."""

    AUTO = 0
    """Auto White Balance"""

    MANUAL = 1
    """Manual White Balance"""


ATTRIBUTE_ID_MAP = {
    # TIFF Rev. 6.0 Section A: Image Data Structure Tags
    "image_width": 256,
    "image_height": 257,
    "bits_per_sample": 258,
    "compression": 259,
    "photometric_interpretation": 262,
    "orientation": 274,
    "samples_per_pixel": 277,
    "planar_configuration": 284,
    "subsampling_ratio_of_y_to_c": 530,
    "y_and_c_positioning": 531,
    "x_resolution": 282,
    "y_resolution": 283,
    "resolution_unit": 296,
    # TIFF Rev. 6.0 Section B: Recording Offset Tags
    "strip_offsets": 273,
    "rows_per_strip": 278,
    "strip_byte_counts": 279,
    "jpeg_interchange_format": 513,
    "jpeg_interchange_format_length": 514,
    # TIFF Rev. 6.0 Section C: Image Data Characteristic Tags
    "transfer_function": 301,
    "white_point": 318,
    "primary_chromaticities": 319,
    "matrix_coefficients": 529,
    "reference_black_white": 532,
    # TIFF Rev. 6.0 Section D: Other Tags
    "datetime": 306,
    "image_description": 270,
    "make": 271,
    "model": 272,
    "software": 305,
    "artist": 315,
    "copyright": 33432,
    # EXIF Tags
    "exposure_time": 33434,
    "f_number": 33437,
    "exposure_program": 34850,
    "spectral_sensitivity": 34852,
    "photographic_sensitivity": 34855,
    "oecf": 34856,
    "sensitivity_type": 34864,
    "standard_output_sensitivity": 34865,
    "recommended_exposure_index": 34866,
    "iso_speed": 34867,
    "iso_speed_latitude_yyy": 34868,
    "iso_speed_latitude_zzz": 34869,
    "exif_version": 36864,
    "datetime_original": 36867,
    "datetime_digitized": 36868,
    "offset_time": 36880,
    "offset_time_original": 36881,
    "offset_time_digitized": 36882,
    "components_configuration": 37121,
    "compressed_bits_per_pixel": 37122,
    "shutter_speed_value": 37377,
    "aperture_value": 37378,
    "brightness_value": 37379,
    "exposure_bias_value": 37380,
    "max_aperture_value": 37381,
    "subject_distance": 37382,
    "metering_mode": 37383,
    "light_source": 37384,
    "flash": 37385,
    "focal_length": 37386,
    "subject_area": 37396,
    "maker_note": 37500,
    "user_comment": 37510,
    "subsec_time": 37520,
    "subsec_time_original": 37521,
    "subsec_time_digitized": 37522,
    "temperature": 37888,
    "humidity": 37889,
    "pressure": 37890,
    "water_depth": 37891,
    "acceleration": 37892,
    "camera_elevation_angle": 37893,
    "xp_title": 0x9c9b,
    "xp_comment": 0x9c9c,
    "xp_author": 0x9c9d,
    "xp_keywords": 0x9c9e,
    "xp_subject": 0x9c9f,
    "flashpix_version": 40960,
    "color_space": 40961,
    "pixel_x_dimension": 40962,
    "pixel_y_dimension": 40963,
    "related_sound_file": 40964,
    "flash_energy": 41483,
    "spatial_frequency_response": 41484,
    "focal_plane_x_resolution": 41486,
    "focal_plane_y_resolution": 41487,
    "focal_plane_resolution_unit": 41488,
    "subject_location": 41492,
    "exposure_index": 41493,
    "sensing_method": 41495,
    "file_source": 41728,
    "scene_type": 41729,
    "cfa_pattern": 41730,
    "custom_rendered": 41985,
    "exposure_mode": 41986,
    "white_balance": 41987,
    "digital_zoom_ratio": 41988,
    "focal_length_in_35mm_film": 41989,
    "scene_capture_type": 41990,
    "gain_control": 41991,
    "contrast": 41992,
    "saturation": 41993,
    "sharpness": 41994,
    "device_setting_description": 41995,
    "subject_distance_range": 41996,
    "image_unique_id": 42016,
    "camera_owner_name": 42032,
    "body_serial_number": 42033,
    "lens_specification": 42034,
    "lens_make": 42035,
    "lens_model": 42036,
    "lens_serial_number": 42037,
    "gamma": 42240,
    # GPS Info Tags
    "gps_version_id": 0,
    "gps_latitude_ref": 1,
    "gps_latitude": 2,
    "gps_longitude_ref": 3,
    "gps_longitude": 4,
    "gps_altitude_ref": 5,
    "gps_altitude": 6,
    "gps_timestamp": 7,
    "gps_satellites": 8,
    "gps_status": 9,
    "gps_measure_mode": 10,
    "gps_dop": 11,
    "gps_speed_ref": 12,
    "gps_speed": 13,
    "gps_track_ref": 14,
    "gps_track": 15,
    "gps_img_direction_ref": 16,
    "gps_img_direction": 17,
    "gps_map_datum": 18,
    "gps_dest_latitude_ref": 19,
    "gps_dest_latitude": 20,
    "gps_dest_longitude_ref": 21,
    "gps_dest_longitude": 22,
    "gps_dest_bearing_ref": 23,
    "gps_dest_bearing": 24,
    "gps_dest_distance_ref": 25,
    "gps_dest_distance": 26,
    "gps_processing_method": 27,
    "gps_area_information": 28,
    "gps_datestamp": 29,
    "gps_differential": 30,
    "gps_horizontal_positioning_error": 31,
    # EXIF-specific IFDs (for __dir__ Display)
    "_exif_ifd_pointer": 34665,
    "_gps_ifd_pointer": 34853,
    "_interoperability_ifd_Pointer": 40965,
}

ATTRIBUTE_NAME_MAP = {value: key for key, value in ATTRIBUTE_ID_MAP.items()}

ATTRIBUTE_TYPE_MAP = {  # tuple of type ID and IFD number description used when adding new tags
    "aperture_value": (int(ExifTypes.RATIONAL), "exif"),
    "artist": (int(ExifTypes.ASCII), 0),
    "brightness_value": (int(ExifTypes.SRATIONAL), "exif"),
    "color_space": (int(ExifTypes.SHORT), "exif"),
    "contrast": (int(ExifTypes.SHORT), "exif"),
    "copyright": (int(ExifTypes.ASCII), 0),
    "custom_rendered": (int(ExifTypes.SHORT), "exif"),
    "datetime": (int(ExifTypes.ASCII), 0),
    "datetime_digitized": (int(ExifTypes.ASCII), "exif"),
    "datetime_original": (int(ExifTypes.ASCII), "exif"),
    "digital_zoom_ratio": (int(ExifTypes.RATIONAL), "exif"),
    "exposure_bias_value": (int(ExifTypes.SRATIONAL), "exif"),
    "exposure_index": (int(ExifTypes.RATIONAL), "exif"),
    "exposure_mode": (int(ExifTypes.SHORT), "exif"),
    "exposure_program": (int(ExifTypes.SHORT), "exif"),
    "exposure_time": (int(ExifTypes.RATIONAL), "exif"),
    "f_number": (int(ExifTypes.RATIONAL), "exif"),
    "flash": (int(ExifTypes.SHORT), "exif"),
    "flash_energy": (int(ExifTypes.RATIONAL), "exif"),
    "focal_length_in_35mm_film": (int(ExifTypes.SHORT), "exif"),
    "focal_plane_resolution_unit": (int(ExifTypes.SHORT), "exif"),
    "gain_control": (int(ExifTypes.RATIONAL), "exif"),
    "gps_altitude": (int(ExifTypes.RATIONAL), "gps"),
    "gps_altitude_ref": (int(ExifTypes.BYTE), "gps"),
    "gps_datestamp": (int(ExifTypes.ASCII), "gps"),
    "gps_dest_bearing": (int(ExifTypes.RATIONAL), "gps"),
    "gps_dest_bearing_ref": (int(ExifTypes.ASCII), "gps"),
    "gps_dest_distance": (int(ExifTypes.RATIONAL), "gps"),
    "gps_dest_distance_ref": (int(ExifTypes.ASCII), "gps"),
    "gps_dest_latitude": (int(ExifTypes.RATIONAL), "gps"),
    "gps_dest_latitude_ref": (int(ExifTypes.ASCII), "gps"),
    "gps_dest_longitude": (int(ExifTypes.RATIONAL), "gps"),
    "gps_dest_longitude_ref": (int(ExifTypes.ASCII), "gps"),
    "gps_differential": (int(ExifTypes.SHORT), "gps"),
    "gps_dop": (int(ExifTypes.RATIONAL), "gps"),
    "gps_img_direction": (int(ExifTypes.RATIONAL), "gps"),
    "gps_img_direction_ref": (int(ExifTypes.ASCII), "gps"),
    "gps_latitude": (int(ExifTypes.RATIONAL), "gps"),
    "gps_latitude_ref": (int(ExifTypes.ASCII), "gps"),
    "gps_longitude": (int(ExifTypes.RATIONAL), "gps"),
    "gps_longitude_ref": (int(ExifTypes.ASCII), "gps"),
    "gps_map_datum": (int(ExifTypes.ASCII), "gps"),
    "gps_measure_mode": (int(ExifTypes.ASCII), "gps"),
    "gps_satellites": (int(ExifTypes.ASCII), "gps"),
    "gps_speed": (int(ExifTypes.RATIONAL), "gps"),
    "gps_speed_ref": (int(ExifTypes.ASCII), "gps"),
    "gps_status": (int(ExifTypes.ASCII), "gps"),
    "gps_timestamp": (int(ExifTypes.RATIONAL), "gps"),
    "gps_track": (int(ExifTypes.RATIONAL), "gps"),
    "gps_track_ref": (int(ExifTypes.ASCII), "gps"),
    "gps_version_id": (int(ExifTypes.BYTE), "gps"),
    "image_description": (int(ExifTypes.ASCII), 0),
    "image_unique_id": (int(ExifTypes.ASCII), "exif"),
    "light_source": (int(ExifTypes.SHORT), "exif"),
    "make": (int(ExifTypes.ASCII), 0),
    "max_aperture_value": (int(ExifTypes.RATIONAL), 0),
    "metering_mode": (int(ExifTypes.SHORT), "exif"),
    "model": (int(ExifTypes.ASCII), 0),
    "orientation": (int(ExifTypes.SHORT), 0),
    "pixel_x_dimension": (int(ExifTypes.SHORT), "exif"),
    "pixel_y_dimension": (int(ExifTypes.SHORT), "exif"),
    "saturation": (int(ExifTypes.SHORT), "exif"),
    "scene_capture_type": (int(ExifTypes.SHORT), "exif"),
    "sensing_method": (int(ExifTypes.SHORT), "exif"),
    "shutter_speed_value": (int(ExifTypes.SRATIONAL), "exif"),
    "software": (int(ExifTypes.ASCII), 0),
    "sharpness": (int(ExifTypes.SHORT), "exif"),
    "spectral_sensitivity": (int(ExifTypes.ASCII), "exif"),
    "subsec_time": (int(ExifTypes.ASCII), "exif"),
    "subsec_time_original": (int(ExifTypes.ASCII), "exif"),
    "subsec_time_digitized": (int(ExifTypes.ASCII), "exif"),
    "subject_distance": (int(ExifTypes.RATIONAL), "exif"),
    "subject_distance_range": (int(ExifTypes.SHORT), "exif"),
    "subject_location": (int(ExifTypes.SHORT), "exif"),
    "white_balance": (int(ExifTypes.SHORT), "exif"),
    "_exif_ifd_pointer": (int(ExifTypes.LONG), 0),
    "_gps_ifd_pointer": (int(ExifTypes.LONG), 0),
}


ERROR_IMG_NO_ATTR = "image does not have attribute {0}"
